tests/kani/write_build_ninja.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+from pathlib import Path
+import re
+import subprocess
+import sys
+
+import ninja_syntax
+
+RULE_NAME = 'kani'
+BUILD_COUNT = 0
+
+def rust_sources(root):
+    sources = subprocess.run(
+        ['find', root, '-name', '*.rs'],
+        encoding='utf-8',
+        universal_newlines=True,
+        check=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE
+    ).stdout.splitlines()
+    # source files with names including 'ignore' and 'fixme' are broken
+    # source file 'ForeignItems/main.rs' is also broken
+    tags = ['ignore', 'fixme', 'tests/kani/ForeignItems/main.rs']
+    return [src for src in sources if all(tag not in src for tag in tags)]
+
+def kani_pragma(pragma, rust_code):
+    # Examples of kani pragmas:
+    # // kani-verify-fail
+    # // kani-check-fail
+    # // kani-flags: --default-unwind 8
+    try:
+        regexp = fr'//[ \t]*{pragma}([^\n]*)'
+        return re.search(regexp, rust_code).group(1).strip()
+    except AttributeError: # 'NoneType' object has no attribute 'group'
+        return None
+
+def kani_verify_fail(rust_code):
+    return kani_pragma('kani-verify-fail', rust_code) is not None
+
+def kani_check_fail(rust_code):
+    return kani_pragma('kani-check-fail', rust_code) is not None
+
+def kani_flags(rust_code):
+    return kani_pragma('kani-flags:', rust_code) or ""
+
+def ninja_build_statement(source):
+    source = Path(source)
+    with open(source, encoding='utf-8') as src:
+        rust_code = src.read()
+    if kani_verify_fail(rust_code) or kani_check_fail(rust_code):
+        return None
+
+    # Warning: concurrent invocations of kani in a single directory
+    # should generate race conditions that have so far been
+    # unobserved.  The count variable here is just a hack just to get
+    # ninja to run.
```

Comment:
So use a pool?

Reply:
That would require one pool per directory.  I'm going to leave this now and add this as a TODO in the code.

Reply:
I implemented the pools, it wasn't that hard.